x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java

```diff
@@ -983,6 +984,39 @@ public LogicalPlan apply(LogicalPlan p) {
         }
     }
 
+    // This class is a workaround for the SUM(all zeros) = NULL issue raised in https://github.com/elastic/elasticsearch/issues/45251 and
+    // should be removed as soon as root cause is fixed and the sum aggregation results can differentiate between SUM(all zeroes) 
+    // and SUM(all nulls)
+    // NOTE: this rule should always be applied AFTER the ReplaceAggsWithStats rule
+    static class ReplaceSumWithStats extends OptimizerBasicRule {
+        
+        @Override 
+        public LogicalPlan apply(LogicalPlan plan) {
+            final Map<Expression, Stats> statsPerField = new LinkedHashMap<>();
+            
+            plan.forEachExpressionsUp(e -> {
```

Comment:
Not related to this PR, but I was wondering: most `forEachExpressionsUp/Down` methods invocations do pattern matching as first thing. Wouldn't an alternative method similar to `Node#forEachUp/Down` taking a type token make sense?

Reply:
The issue lies with collections. Expressions are not just used as individual nodes but also as properties. Take `Project(List<NamedExpression> projections)` - this led to issues in not only filtering but in reconstructing said collections with the new expressions while preserving their types. See the comment in `LogicalPlan.doTransformExpression`

It would be nicer to do:
`plan.forEachExpressionsUp(s -> , Sum.class)` instead of doing the `instanceof` check however the issue right now is preserving the type information before and after transformation without causing a CCE.

That said, I plan to take another look at this to see whether it can be sorted out.